README.md


<a href="https://github.com/SankethBK/flutter_sticky_widgets/actions"><img src="https://github.com/SankethBK/flutter_sticky_widgets/workflows/test-flutter-sticky-widgets/badge.svg" alt="Build Status"></a>


With *flutter_sticky_widgets* you can use a widget as if it is `position: sticky;` in CSS. 

Sticky widgets is an UI pattern where we want widgets to scroll along with the UI upto a certain point, and then stick to the screen instead of scrolling off the screen.

_(Green play button in first image, and shuffle play button in second image are examples for position: sticky)_

<div>
        <img src="https://user-images.githubusercontent.com/51091231/199099550-ae2a8cd2-cab6-42a7-9fcf-078522160e91.gif"> 
        <img src="https://user-images.githubusercontent.com/51091231/199097900-e51e7db2-81e3-409f-a431-16d5baad0f0f.gif">
</div>

## Getting started

```dart
import 'package:flutter_sticky_widgets/flutter_sticky_widgets.dart';
```

```dart
StickyContainer(
    stickyChildren: [
        StickyWidget(
            initialPosition: StickyPosition(bottom: 20, right: 20),
            finalPosition: StickyPosition(
                bottom: MediaQuery.of(context).size.height - 80, right: 20),
            controller: _controller,
            child: Container(
            width: 40,
            height: 40,
            decoration: const BoxDecoration(
                shape: BoxShape.circle,
                color: Colors.green,
            ),
            child: const Center(child: Icon(Icons.pause)),
            ),
        ),
    ],
    child: ListView.builder(
    controller: _controller,
    itemCount: 50,
    itemBuilder: (context, index) {
        return ListTile(
            title: Text("Tile $index"),
        );
    }),
);
```

#### StickyWidget

Widget used to implement the sticky pattern. 

* __initialPosition (StickyPosition):__ indicates the position of the widget before the scrolling has started. Position is represented by a custom class *StickyPosition*. 

* __finalPosition (StickyPosition):__ indicates the position of the widget where it will remain sticky after sufficient amount of scrolling. Widget will return to its initial position if scrolled in reverse direction. 

* __controller (ScrollController):__ used by StickyWidget to detect scrolling. Same ScrollController must be attached to your *Scrollable* (ListView, SingleChildScrollView, CustomScrollView, ...) for StickyWidgets to work. It is also used to determine if the scroll axis is vertical or horizontal. 

* __callback (Function(double)?) (optional):__ An optional callback function, which gets called on every scroll. It receives *scrolloffset* as parameter which can be used to animate the *StickyWidget*. 

#### StickyContainer

It must be the parent element of the *Scrollable* widget used. StickyContainer's size is used as reference size by StickWidgets to position themselves. So make sure StickyContainer's size is same as the size of the *Scrollable* widget. 

* __child (Widget):__ ideally *Scrollable* widget should be supplied for this parameter, if not you have to make sure StickyContainer's size is same as *Scrollable's* size.

* __stickyChildren (List\<StickyWidget>):__ you can supply one or more *StickyWidgets*.

* __displayOverFlowContent (bool) (optional):__ true if you want to display the StickyWidgets overflowing outside of the StickyContainer's bounds. 
 
#### StickyPosition

A Custom class used to represent initial and final positions of *StickyWidget*. It takes 4 parameters (top, bottom, left, right). But you should supply only 2 of them (top / bottom, right / left)

Note: `Make sure to pass same set of parameters for StickyPosition used in initialPosition and finalPosition`

Eg: 

Correct ✅  
initialPosition: StickyPosition(top: .., left: ..)  
finalPosition: StickyPosition(top: .., left: ..) 

Wrong ❌  
(should supply top instead of bottom for finalPosition also)


initialPosition: StickyPosition(top: .., left: ..)  
finalPosition: StickyPosition(bottom: .., left: ..) 



CHANGELOG.md
## 0.0.1

*  Initial release

## 0.0.2

*  Updated readme

## 0.0.3

*  Added Dart doc comments

test/vertical_list_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_sticky_widgets/flutter_sticky_widgets.dart';

import 'widgets/vertical_list.dart';

void main() {
  Widget buildApp(
      StickyPosition initialPosition, StickyPosition finalScrollPosition) {
    return VerticalList(
        initialPosition: initialPosition, finalPosition: finalScrollPosition);
  }

  group("Test vertical list with bottom position", () {
    testWidgets("Full scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(bottom: 0, right: 20),
          StickyPosition(bottom: 300, right: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item

      final listFinder = find.byType(Scrollable);
      final itemFinder = find.text("item 29");

      await tester.scrollUntilVisible(
        itemFinder,
        100.0,
        scrollable: listFinder,
      );

      expect(find.text("300.0"), findsOneWidget);
    });

    testWidgets("Half scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(bottom: 0, right: 20),
          StickyPosition(bottom: 300, right: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item
      final listView = tester.widget<ListView>(find.byType(ListView));
      final ctrl = listView.controller;

      ctrl?.jumpTo(ctrl.offset + 150);
      await tester.pump();

      expect(find.text("150.0"), findsOneWidget);
    });

    testWidgets(
        "Should not immediately scroll down after reaching top and list is scrolled down",
        (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(bottom: 0, right: 20),
          StickyPosition(bottom: 300, right: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll down by 350 px

      await tester.drag(find.byType(ListView), const Offset(0.0, -350.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 25px. sticky widget still expected to be at 325px

      await tester.drag(find.byType(ListView), const Offset(0.0, 25.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 75px. sticky widget  expected to be at 275px

      await tester.drag(find.byType(ListView), const Offset(0.0, 50.0));
      await tester.pump();

      expect(find.text("275.0"), findsOneWidget);
    });
  });

  group("Test vertical list with top position", () {
    testWidgets("Full scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(top: 0, right: 20),
          StickyPosition(top: 300, right: 20)));
      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item

      final listFinder = find.byType(Scrollable);
      final itemFinder = find.text("item 29");

      await tester.scrollUntilVisible(
        itemFinder,
        100.0,
        scrollable: listFinder,
      );

      expect(find.text("300.0"), findsOneWidget);
    });

    testWidgets("Half scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(bottom: 0, right: 20),
          StickyPosition(bottom: 300, right: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item
      final listView = tester.widget<ListView>(find.byType(ListView));
      final ctrl = listView.controller;

      ctrl?.jumpTo(ctrl.offset + 150);
      await tester.pump();

      expect(find.text("150.0"), findsOneWidget);
    });

    testWidgets(
        "Should not immediately scroll down after reaching top and list is scrolled down",
        (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(bottom: 0, right: 20),
          StickyPosition(bottom: 300, right: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll down by 350 px

      await tester.drag(find.byType(ListView), const Offset(0.0, -350.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 25px. sticky widget still expected to be at 325px

      await tester.drag(find.byType(ListView), const Offset(0.0, 25.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 75px. sticky widget  expected to be at 275px

      await tester.drag(find.byType(ListView), const Offset(0.0, 50.0));
      await tester.pump();

      expect(find.text("275.0"), findsOneWidget);
    });
  });
}


test/horizontal_list_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_sticky_widgets/flutter_sticky_widgets.dart';

import 'widgets/horizontal_list.dart';

void main() {
  Widget buildApp(
      StickyPosition initialPosition, StickyPosition finalScrollPosition) {
    return HorizontalList(
        initialPosition: initialPosition, finalPosition: finalScrollPosition);
  }

  group("Test horizontal list with right position", () {
    testWidgets("Full scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(right: 0, bottom: 20),
          StickyPosition(right: 300, bottom: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item

      final listFinder = find.byType(Scrollable);
      final itemFinder = find.text("item 29");

      await tester.scrollUntilVisible(
        itemFinder,
        100.0,
        scrollable: listFinder,
      );

      expect(find.text("300.0"), findsOneWidget);
    });

    testWidgets("Half scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(right: 0, bottom: 20),
          StickyPosition(right: 300, bottom: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item
      final listView = tester.widget<ListView>(find.byType(ListView));
      final ctrl = listView.controller;

      ctrl?.jumpTo(ctrl.offset + 150);
      await tester.pump();

      expect(find.text("150.0"), findsOneWidget);
    });

    testWidgets(
        "Should not immediately scroll left after reaching right and list is scrolled left",
        (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(right: 0, bottom: 20),
          StickyPosition(right: 300, bottom: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll down by 350 px

      await tester.drag(find.byType(ListView), const Offset(-350.0, 0.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 25px. sticky widget still expected to be at 325px

      await tester.drag(find.byType(ListView), const Offset(25.0, 0.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 75px. sticky widget  expected to be at 275px

      await tester.drag(find.byType(ListView), const Offset(50.0, 0.0));
      await tester.pump();

      expect(find.text("275.0"), findsOneWidget);
    });
  });

  group("Test horizontal list with top position", () {
    testWidgets("Full scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(left: 0, bottom: 20),
          StickyPosition(left: 300, bottom: 20)));
      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item

      final listFinder = find.byType(Scrollable);
      final itemFinder = find.text("item 29");

      await tester.scrollUntilVisible(
        itemFinder,
        100.0,
        scrollable: listFinder,
      );

      expect(find.text("300.0"), findsOneWidget);
    });

    testWidgets("Half scroll", (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(left: 0, bottom: 20),
          StickyPosition(left: 300, bottom: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll till the last item
      final listView = tester.widget<ListView>(find.byType(ListView));
      final ctrl = listView.controller;

      ctrl?.jumpTo(ctrl.offset + 150);
      await tester.pump();

      expect(find.text("150.0"), findsOneWidget);
    });

    testWidgets(
        "Should not immediately scroll down after reaching top and list is scrolled down",
        (WidgetTester tester) async {
      await tester.pumpWidget(buildApp(StickyPosition(left: 0, bottom: 20),
          StickyPosition(left: 300, bottom: 20)));

      // initial position
      expect(find.text("0.0"), findsOneWidget);
      expect(find.byType(Scrollable), findsOneWidget);

      // scroll down by 350 px

      await tester.drag(find.byType(ListView), const Offset(-350.0, 0.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 25px. sticky widget still expected to be at 325px

      await tester.drag(find.byType(ListView), const Offset(25.0, 0.0));
      await tester.pump();

      expect(find.text("300.0"), findsOneWidget);

      // scroll up by 75px. sticky widget  expected to be at 275px

      await tester.drag(find.byType(ListView), const Offset(50.0, 0.0));
      await tester.pump();

      expect(find.text("275.0"), findsOneWidget);
    });
  });
}


test/widgets/vertical_list.dart
import 'package:flutter/material.dart';
import 'package:flutter_sticky_widgets/flutter_sticky_widgets.dart';

class VerticalList extends StatefulWidget {
  final StickyPosition initialPosition;
  final StickyPosition finalPosition;
  const VerticalList(
      {Key? key, required this.initialPosition, required this.finalPosition})
      : super(key: key);

  @override
  State<VerticalList> createState() => _VerticalListState();
}

class _VerticalListState extends State<VerticalList> {
  double buttonText = 0;

  late ScrollController controller;

  @override
  void initState() {
    controller = ScrollController();
    super.initState();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: SizedBox(
          height: 300,
          child: StickyContainer(
            stickyChildren: [
              StickyWidget(
                callback: (double scrollffset) => setState(() {
                  buttonText = scrollffset;
                }),
                initialPosition: widget.initialPosition,
                finalPosition: widget.finalPosition,
                controller: controller,
                child: Container(
                  width: 40,
                  height: 40,
                  decoration: const BoxDecoration(
                      shape: BoxShape.circle, color: Colors.red),
                  child: Center(child: Text("$buttonText")),
                ),
              )
            ],
            child: ListView.builder(
              itemCount: 30,
              controller: controller,
              itemBuilder: (context, index) => SizedBox(
                height: 30,
                child: Text("item $index"),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


lib/src/sticky_widget.dart
import 'package:flutter/material.dart';
import 'sticky_position.dart';

class StickyWidget extends StatefulWidget {
  /// indicates the position of the widget before the scrolling has started.
  final StickyPosition _initialPosition;

  /// indicates the position of the widget where it will remain sticky after sufficient amount of scrolling.
  final StickyPosition _finalPosition;

  /// used by StickyWidget to detect scrolling. Same ScrollController must be attached to your Scrollable.
  final ScrollController _controller;

  /// An optional callback function, which gets called on every scroll. It receives scrolloffset as parameter which can be used to animate the StickyWidget.
  final Function(double)? callback;
  final Widget child;
  StickyWidget(
      {Key? key,
      required StickyPosition initialPosition,
      required StickyPosition finalPosition,
      required ScrollController controller,
      required this.child,
      this.callback})
      : assert(
            !((initialPosition.top != null) ^ (finalPosition.top != null)) &&
                !((initialPosition.bottom != null) ^
                    (finalPosition.bottom != null)) &&
                !((initialPosition.left != null) ^
                    (finalPosition.left != null)) &&
                !((initialPosition.left != null) ^
                    (finalPosition.left != null)),
            "Provide same position paramters for initial and final Positions"),
        assert(
            (initialPosition.top != null) ^ (initialPosition.bottom != null) &&
                (initialPosition.left != null) ^
                    (initialPosition.right != null) &&
                (finalPosition.top != null) ^ (finalPosition.bottom != null) &&
                (finalPosition.left != null) ^ (finalPosition.right != null),
            "Don't provide both parameters of same axis (eg: top, bottom)"),
        _initialPosition = initialPosition,
        _finalPosition = finalPosition,
        _controller = controller,
        super(key: key);

  @override
  State<StickyWidget> createState() => _StickyWidgetState();
}

class _StickyWidgetState extends State<StickyWidget> {
  // Only 2 of these will be used, depending on which one is supplied by the user
  double? _computedTop;
  double? _computedBottom;
  double? _computedLeft;
  double? _computedRight;
  late Axis _scrollAxis;

  @override
  void initState() {
    _computedTop = widget._initialPosition.top;
    _computedBottom = widget._initialPosition.bottom;
    _computedLeft = widget._initialPosition.left;
    _computedRight = widget._initialPosition.right;
    _scrollAxis = widget._controller.position.axis;

    widget._controller.addListener(updatePosition);
    super.initState();
  }

  @override
  void dispose() {
    widget._controller.removeListener((updatePosition));
    super.dispose();
  }

  void updatePosition() {
    if (_scrollAxis == Axis.vertical) {
      if (widget._initialPosition.top != null) {
        updateTop(widget._controller.position.pixels);
      } else {
        updateBottom(widget._controller.position.pixels);
      }
    } else {
      if (widget._initialPosition.left != null) {
        updateLeft(widget._controller.position.pixels);
      } else {
        updateRight(widget._controller.position.pixels);
      }
    }
  }

  void updateTop(double scrollOffset) {
    double finalTop;
    if (widget._initialPosition.top! < widget._finalPosition.top!) {
      finalTop = widget._initialPosition.top! + scrollOffset;

      if (finalTop >= widget._initialPosition.top! &&
          finalTop <= widget._finalPosition.top!) {
        setState(() {
          _computedTop = finalTop;
          if (widget.callback != null) widget.callback!(_computedTop!);
        });
      } else if (finalTop > widget._finalPosition.top! &&
          _computedTop! < widget._finalPosition.top!) {
        setState(() {
          _computedTop = widget._finalPosition.top!;
          if (widget.callback != null) widget.callback!(_computedTop!);
        });
      }
    } else {
      finalTop = widget._initialPosition.top! - scrollOffset;

      if (finalTop <= widget._initialPosition.top! &&
          finalTop >= widget._finalPosition.top!) {
        setState(() {
          _computedTop = finalTop;
          if (widget.callback != null) widget.callback!(_computedTop!);
        });
      } else if (finalTop < widget._finalPosition.top! &&
          _computedTop! > widget._finalPosition.top!) {
        setState(() {
          _computedTop = widget._finalPosition.top!;
          if (widget.callback != null) widget.callback!(_computedTop!);
        });
      }
    }
  }

  void updateBottom(double scrollOffset) {
    double finalBottom;

    if (widget._initialPosition.bottom! < widget._finalPosition.bottom!) {
      finalBottom = widget._initialPosition.bottom! + scrollOffset;

      if (finalBottom >= widget._initialPosition.bottom! &&
          finalBottom <= widget._finalPosition.bottom!) {
        setState(() {
          _computedBottom = finalBottom;
          if (widget.callback != null) widget.callback!(_computedBottom!);
        });
      } else if (finalBottom > widget._finalPosition.bottom! &&
          _computedBottom! < widget._finalPosition.bottom!) {
        setState(() {
          _computedBottom = widget._finalPosition.bottom!;
          if (widget.callback != null) widget.callback!(_computedBottom!);
        });
      }
    } else {
      finalBottom = widget._initialPosition.bottom! - scrollOffset;

      if (finalBottom <= widget._initialPosition.bottom! &&
          finalBottom >= widget._finalPosition.bottom!) {
        setState(() {
          _computedBottom = finalBottom;
          if (widget.callback != null) widget.callback!(_computedBottom!);
        });
      } else if (finalBottom < widget._finalPosition.bottom! &&
          _computedBottom! > widget._finalPosition.bottom!) {
        setState(() {
          _computedBottom = widget._finalPosition.bottom!;
          if (widget.callback != null) widget.callback!(_computedBottom!);
        });
      }
    }
  }

  void updateLeft(double scrollOffset) {
    double finalLeft;
    if (widget._initialPosition.left! < widget._finalPosition.left!) {
      finalLeft = widget._initialPosition.left! + scrollOffset;

      if (finalLeft >= widget._initialPosition.left! &&
          finalLeft <= widget._finalPosition.left!) {
        setState(() {
          _computedLeft = finalLeft;
          if (widget.callback != null) widget.callback!(_computedLeft!);
        });
      } else if (finalLeft > widget._finalPosition.left! &&
          _computedLeft! < widget._finalPosition.left!) {
        setState(() {
          _computedLeft = widget._finalPosition.left!;
          if (widget.callback != null) widget.callback!(_computedLeft!);
        });
      }
    } else {
      finalLeft = widget._initialPosition.left! - scrollOffset;

      if (finalLeft <= widget._initialPosition.left! &&
          finalLeft >= widget._finalPosition.left!) {
        setState(() {
          _computedLeft = finalLeft;
          if (widget.callback != null) widget.callback!(_computedLeft!);
        });
      } else if (finalLeft < widget._finalPosition.left! &&
          _computedLeft! > widget._finalPosition.left!) {
        setState(() {
          _computedLeft = widget._finalPosition.left!;
          if (widget.callback != null) widget.callback!(_computedLeft!);
        });
      }
    }
  }

  void updateRight(double scrollOffset) {
    double finalRight;
    if (widget._initialPosition.right! < widget._finalPosition.right!) {
      finalRight = widget._initialPosition.right! + scrollOffset;

      if (finalRight >= widget._initialPosition.right! &&
          finalRight <= widget._finalPosition.right!) {
        setState(() {
          _computedRight = finalRight;
          if (widget.callback != null) widget.callback!(_computedRight!);
        });
      } else if (finalRight > widget._finalPosition.right! &&
          _computedRight! < widget._finalPosition.right!) {
        setState(() {
          _computedRight = widget._finalPosition.right!;
          if (widget.callback != null) widget.callback!(_computedRight!);
        });
      }
    } else {
      finalRight = widget._initialPosition.right! - scrollOffset;

      if (finalRight <= widget._initialPosition.right! &&
          finalRight >= widget._finalPosition.right!) {
        setState(() {
          _computedRight = finalRight;
          if (widget.callback != null) widget.callback!(_computedRight!);
        });
      } else if (finalRight < widget._finalPosition.right! &&
          _computedRight! > widget._finalPosition.right!) {
        setState(() {
          _computedRight = widget._finalPosition.right!;
          if (widget.callback != null) widget.callback!(_computedRight!);
        });
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return Positioned(
      top: _computedTop,
      bottom: _computedBottom,
      left: _computedLeft,
      right: _computedRight,
      child: widget.child,
    );
  }
}


lib/src/sticky_container.dart
import 'package:flutter/material.dart';
import 'package:flutter_sticky_widgets/flutter_sticky_widgets.dart';

/// Widget to be around StickyWidgets,
///
/// StickyWidgets use the size of its parent StickyContainer as reference for positioning
class StickyContainer extends StatelessWidget {
  /// Ideally a Scrollable widget should be supplied for this parameter
  final Widget child;

  /// You can pass a list of StickyWidgets, StickyContainer does not ensures they are non-overlapping
  final List<StickyWidget> stickyChildren;

  /// true if you want to display the StickyWidgets overflowing outside of the StickyContainer's bounds.
  final bool displayOverFlowContent;
  const StickyContainer(
      {Key? key,
      required this.child,
      required this.stickyChildren,
      this.displayOverFlowContent = false})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      clipBehavior: Clip.none,
      children: [
        child,
        ...stickyChildren,
      ],
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_sticky_widgets/flutter_sticky_widgets.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late ScrollController _controller;

  @override
  void initState() {
    _controller = ScrollController();
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: StickyContainer(
        displayOverFlowContent: true,
        stickyChildren: [
          StickyWidget(
            initialPosition: StickyPosition(bottom: 20, right: 20),
            finalPosition: StickyPosition(
                bottom: MediaQuery.of(context).size.height - 80, right: 20),
            controller: _controller,
            child: Container(
              width: 40,
              height: 40,
              decoration: const BoxDecoration(
                shape: BoxShape.circle,
                color: Colors.green,
              ),
              child: const Center(child: Icon(Icons.pause)),
            ),
          ),
        ],
        child: ListView.builder(
            controller: _controller,
            itemCount: 50,
            itemBuilder: ((context, index) {
              return ListTile(
                title: Text("Tile $index"),
              );
            })),
      ),
    );
  }
}
